### PR TITLE
don't set HOME on Windows

### LINKED ISSUE #1946: [RFC] unset HOME environment variable when spawn new process on Windows
Current implementation of Windows version set HOME enviroment variable to USERPROFILE if $HOME is not set on start-up. But there are some applications what doesn't expect this behavior. For example, msys2/cygwin is that. They have own $HOME environment variable and the place of $HOME not same to USERPROFILE. When we want to spawn new process from vim, it make problems. So I suggest to solve this issue.

## 1. don't set HOME

The most desirable fix is this to me.

## 2. unset HOME before spawning new processes.

But this is only when we have not set HOME before starting vim. If we have set HOME before starting vim, it should be passed into the new process.

I want to hear opinions from users and plugin developers.

### PR DESCRIPTION
**Pull request by mattn**
close #1946

### CODE REVIEW

**Comment by brammool** [comment]
Yasuhiro Matsumoto wrote:

> close #1946

The patch appears to remove code that was needed.
A better explanation is required.


-- 
hundred-and-one symptoms of being an internet addict:
5. You find yourself brainstorming for new subjects to search.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by mattn** [comment]
I reverted removing code for workaround.

**Comment by mattn** [comment]
Should I also update the help file `$HOME-windows` ?

**Comment by brammool** [comment]
Yasuhiro Matsumoto wrote:

> Should I also update the help file `$HOME-windows` ?

I thought I mentioned before: I would prefer keeping the code to set
$HOME, but then not pass on $HOME to an external program.  Thus we need
to remember whether $HOME was set in Vim.

-- 
"So this is it," said Arthur, "we are going to die."
"Yes," said Ford, "except...no!  Wait a minute!"  He suddenly lunged across
the chamber at something behind Arthur's line of vision.  "What's this
switch?" he cried.
"What?   Where?" cried Arthur, twisting around.
"No, I was only fooling," said Ford, "we are going to die after all."
		-- Douglas Adams, "The Hitchhiker's Guide to the Galaxy"

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by mattn** [comment]
I don't understand why you want it. I believe that vim can behave as if it has $HOME.

**Comment by mattn** [comment]
Now, vim behave as if %HOME% has.

Pre-condition  | :echo $HOME       | :echo system('%HOME%')
---------------|-------------------|-----------------------
not set %HOME% | HOME=USERPROFILE  | HOME is not set
HOME=c:/foo    | HOME=c:/foo   | HOME=c:/foo

**Comment by ZyX-I** [comment]
I read @brammool as if he actually wants to keep all code present here, but just remove both `vim_setenv`, and add `$USERPROFILE` support. The tests should be (pseudocode):

```VimL
function CheckHome(exp, ...)
  call assert_equal(a:exp, $HOME)
  call assert_equal(a:exp, expand('~', ':p'))
  if !a:0
    call CheckHomeIsMissingFromSubprocessEnvironment()
  else
    call CheckHomeIsInSubprocessEnvironment(a:exp)
  endif
endfunction
1:
  unlet $HOME $USERPROFILE $HOMEDRIVE $HOMEPATH
  call CheckHome('')
2:
  unlet $HOME $USERPROFILE $HOMEPATH
  let $HOMEDRIVE = 'C:'
  call CheckHome('C:\')
3:
  unlet $HOME $USERPROFILE $HOMEDRIVE
  let $HOMEPATH = 'C'
  call CheckHome('')
4:
  unlet $HOME $USERPROFILE
  let $HOMEDRIVE = 'C:'
  let $HOMEPATH = '\foobar'
  call CheckHome('C:\foobar')
5:
  unlet $HOME $HOMEDRIVE $HOMEPATH
  let $USERPROFILE = 'C:\foo'
  call CheckHome('C:\foo')
6:
  unlet $HOME
  let $USERPROFILE = 'C:\foo'
  let $HOMEDRIVE = 'C:'
  let $HOMEPATH = '\baz'
  call CheckHome('C:\foo')
6:
  let $HOME = 'C:\bar'
  let $USERPROFILE = 'C:\foo'
  let $HOMEDRIVE = 'C:'
  let $HOMEPATH = '\baz'
  call CheckHome('C:\bar', 1)
7:
  let $HOME = '%USERPROFILE%\bar'
  let $USERPROFILE = 'C:\foo'
  let $HOMEDRIVE = 'C:'
  let $HOMEPATH = '\baz'
  call CheckHome('C:\foo\bar', 1)
8:
  let $HOME = '%USERPROFILE'
  let $USERPROFILE = 'C:\foo'
  let $HOMEDRIVE = 'C:'
  let $HOMEPATH = '\baz'
  call CheckHome('%USERPROFILE', 1)
9:
  let $HOME = 'C:\%USERPROFILE%'
  let $USERPROFILE = 'C:\foo'
  let $HOMEDRIVE = 'C:'
  let $HOMEPATH = '\baz'
  call CheckHome('C:\%USERPROFILE%', 1)
```

@brammool Am I right on the test cases?

**Comment by mattn** [comment]
On Windows, we can't set empty value into environment variable.

**Comment by mattn** [comment]
IMO, I hope to change the variable to use get $HOME to USERPROFILE from HOMEDRIVE HOMEPATH.

USERPROFILE is pointed same value to HOMEDRIVE HOMEPATH. And Wikipedia says USERPROFILE is equivalent to HOME on UNIX. (But [here](https://en.wikipedia.org/wiki/Home_directory) says another one)

https://en.wikipedia.org/wiki/Environment_variable#Details

Otherwise, I can add more fallback which get HOMEDRIVE HOMEPATH if USERPROFILE is not set.

**Comment by mattn** [comment]
@ZyX-I Maybe No 7 should be

```
7:
  let $HOME = '%USERPROFILE%\bar'
  let $USERPROFILE = 'C:\foo'
  let $HOMEDRIVE = 'C:'
  let $HOMEPATH = '\baz'
  call CheckHome('%USERPROFILE%\bar', 1)
```
Because the external $HOME should be same as set. Right?

**Comment by mattn** [comment]
fixed to prefer HOMEDRIVE and HOMEPATH. updated tests based on ZyX-I

**Comment by ZyX-I** [comment]
@mattn If I read code correctly No 7 just recorded current behaviour. 8 is checking for false branch of

```C
p = vim_strchr(var + 1, '%');		
if (p != NULL)
```

But why first two lines of CheckHome() were commented, making most of tests almost useless?

File: src/testdir/test_windows_home.vim [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+endfunction
+
+function! TestWindowsHome()
+  command! -nargs=* SaveEnv call <SID>save_env(<f-args>)
+  command! -nargs=* RestoreEnv call <SID>restore_env()
+  command! -nargs=* UnletEnv call <SID>unlet_env(<f-args>)
+
+  SaveEnv $HOME $USERPROFILE $HOMEDRIVE $HOMEPATH
+  try
+    RestoreEnv
```
Comment (ZyX-I) [inline src/testdir/test_windows_home.vim:57]: `RestoreEnv` is useless inside `try`, I have explicitly specified each and every of four values in each test and I see that you copied that.

Comment (mattn) [reply src/testdir/test_windows_home.vim:57]: Thank your pointing this. Are you okay with just removing RestoreEnv in `try`?

**Comment by brammool** [comment]
Nikolai Pavlov wrote:

> @mattn If I read code correctly No 7 just recorded current behaviour. 8 is checking for false branch of
> 
> ```C
> p = vim_strchr(var + 1, '%');		
> if (p != NULL)
> ```
> 
> But why first two lines of CheckHome() were commented, making most of
> tests almost useless?

Actually, the test doesn't do anything, because the function name does
not start with "Test_".

When I fix that the test fails in many ways.  Probably because "homedir"
is set the moment you change $HOME.  Thus that must be done after
changing the others.  Thus the first thing that CheckHome() should do
is:
	let $HOME = ''
Then "homedir" will be recomputed.

-- 
hundred-and-one symptoms of being an internet addict:
31. You code your homework in HTML and give your instructor the URL.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by mattn** [comment]
Sorry. :(

**Pull request closed without merging**